### PR TITLE
feat(s3): support copy/delete/get/head object APIs

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Support S3 client APIs:
* CopyObject
* DeleteObject
* GetObject
* HeadObject

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/AwsClients/S3/getObject.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+				ObjectLockMode: 'x-amz-object-lock-mode',
+				ObjectLockRetainUntilDate: [
+					'x-amz-object-lock-retain-until-date',
+					deserializeTimestamp,
+				],
+				ObjectLockLegalHoldStatus: 'x-amz-object-lock-legal-hold',
+			}),
+			Metadata: deserializeMetadata(response.headers),
+			$metadata: parseMetadata(response),
+			Body: response.body!,
```
Comment (HuiSF) [inline packages/storage/src/AwsClients/S3/getObject.ts:116]: Re: the non-null assertion - does it actually do any validation here?

If the response contains no body (`null`) does it indicate something wrong, and if the assertion fails (response.body === null), should it throw a meaningful exception?

File: packages/storage/src/AwsClients/S3/headObject.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+>;
+
+export type HeadObjectOutput = Pick<
+	HeadObjectCommandOutput,
+	| '$metadata'
+	| 'ContentLength'
+	| 'ContentType'
+	| 'ETag'
+	| 'LastModified'
+	| 'Metadata'
```
Comment (HuiSF) [inline packages/storage/src/AwsClients/S3/headObject.ts:39]: Could include `VersionId`, remembered seeing the ask of this filed in some GH issues.

File: packages/storage/src/AwsClients/S3/utils/deserializeHelpers.ts [hunk 3]
```diff
@@ -116,3 +118,18 @@ export const emptyArrayGuard = <T extends Array<any>>(
...
+/**
+ * @internal
+ */
+export const deserializeMetadata = (
+	headers: Headers
+): Record<string, string> => {
+	const deserialized = Object.keys(headers)
+		.filter(header => header.startsWith('x-amz-meta-'))
+		.reduce((acc, header) => {
+			acc[header.substring(11)] = headers[header];
```
Comment (HuiSF) [inline packages/storage/src/AwsClients/S3/utils/deserializeHelpers.ts:131]: nit:
```typescript
const metadataPrefix = 'x-amz-meta-';
...
acc[header.replace(metadataPrefix, '')]
```

File: packages/storage/src/AwsClients/S3/getObject.ts (hunk 1 above)
Comment (AllanZhengYP) [reply packages/storage/src/AwsClients/S3/getObject.ts:116]: It won't happen as the XHR body defaults to `''` if the request is not straightly errors out. But it never harms to add a check here.

**Review by HuiSF — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

**Pull request merged**